accord-core/src/main/java/accord/utils/async/AsyncChains.java

```diff
@@ -0,0 +1,406 @@
+    // (or perhaps some additional helper implementations that permit us to simply implement apply for Map and FlatMap)
+    <O, T extends AsyncChain<O> & BiConsumer<? super V, Throwable>> AsyncChain<O> add(Function<Head<?>, T> factory)
+    {
+        Preconditions.checkState(next instanceof Head<?>);
+        Head<?> head = (Head<?>) next;
+        T result = factory.apply(head);
+        next = result;
+        return result;
+    }
+
+    <P, O, T extends AsyncChain<O> & BiConsumer<? super V, Throwable>> AsyncChain<O> add(BiFunction<Head<?>, P, T> factory, P param)
+    {
+        Preconditions.checkState(next instanceof Head<?>);
+        Head<?> head = (Head<?>) next;
+        T result = factory.apply(head, param);
+        next = result;
+        return result;
+    }
+
+    private static <V> Runnable encapsulate(Callable<V> callable, BiConsumer<? super V, Throwable> receiver)
+    {
+        return () -> {
+            try
+            {
+                V result = callable.call();
+                receiver.accept(result, null);
+            }
+            catch (Throwable t)
+            {
+                receiver.accept(null, t);
+            }
+        };
+    }
+
+    private static Runnable encapsulate(Runnable runnable, BiConsumer<? super Void, Throwable> receiver)
+    {
+        return () -> {
+            try
+            {
+                runnable.run();
+                receiver.accept(null, null);
+            }
+            catch (Throwable t)
+            {
+                receiver.accept(null, t);
+            }
+        };
+    }
+
+    public static <V> AsyncChain<V> success(V success)
+    {
+        return new Immediate<>(success);
+    }
+
+    public static <V> AsyncChain<V> failure(Throwable failure)
+    {
+        return new Immediate<>(failure);
+    }
+
+    public static <V> AsyncChain<V> ofCallable(Executor executor, Callable<V> callable)
+    {
+        return new Head<V>()
+        {
+            @Override
+            public void begin(BiConsumer<? super V, Throwable> next)
+            {
+                executor.execute(encapsulate(callable, next));
+            }
+        };
+    }
+
+    public static AsyncChain<Void> ofRunnable(Executor executor, Runnable runnable)
+    {
+        return new Head<Void>()
+        {
+            @Override
+            public void begin(BiConsumer<? super Void, Throwable> callback)
+            {
+                executor.execute(AsyncChains.encapsulate(runnable, callback));
+            }
+        };
+    }
+
+    public static <V> AsyncChain<List<V>> all(List<AsyncChain<V>> chains)
+    {
+        Preconditions.checkArgument(!chains.isEmpty());
+        return new AsyncChainCombiner.All<>(chains);
+    }
+
+    public static <V> AsyncChain<V> reduce(List<AsyncChain<V>> chains, BiFunction<V, V, V> reducer)
+    {
+        Preconditions.checkArgument(!chains.isEmpty());
+        if (chains.size() == 1)
+            return chains.get(0);
```

Comment:
this isn't correct behavior, you need a `zero` for this case, else should reject.  There was another bug in accord where `reduce` would change the desired behavior the `reducer` requested when size 1... `reduce` should only ever work with 2+ elements and needs a `zero` to handle single element case.

See `reduce` example in `Stream` `java.util.stream.Stream#reduce(T, java.util.function.BinaryOperator<T>)`; they support single element by returning `Optional` in this case

Reply:
by fixing this case, you can make the return be a different type than `V`, so could have

```
public static <V, R> AsyncChain<R> reduce(V zero, List<AsyncChain<V>> chains, BiFunction<V, V, R> reducer)
```

Though, if you do go that route the below logic makes things complex as types change

```
    if (Reduce.canAppendTo(chains.get(0), reducer))
        {
            AsyncChainCombiner.Reduce<V> appendTo = (AsyncChainCombiner.Reduce<V>) chains.get(0);
            appendTo.addAll(chains.subList(1, chains.size()));
            return appendTo;
        }
```

Reply:
This isn’t a bug, it’s a choice about semantic and I don’t believe there is anything faulty about it? This is sufficient to correctly represent many “reductions”.  It’s not incompatible with a second implementation that accepts a parameter, if we want. That said, I’m not opposed to augmenting this one now, but it’s also fine without AFAICT.

Reply:
Right, this is about the same as `Stream#reduce`, where the reducer is only called if there is more than one result to feed it. The `Optional` returned by `Stream#reduce` is to support empty streams. This is explicitly not supported by this reduce method as you can see by the not-empty check. Everywhere this is currently used, it would be an error to not have at least one chain to reduce. I wouldn't be opposed to adding a zero parameter if that changes though.

Reply:
> Right, this is about the same as Stream#reduce, where the reducer is only called if there is more than one result to feed it.

tested `Stream.of(0).reduce((i1, i2) -> i1 + i2).get();` and it does match your semantics... I think I am remembering scalaz too much here, I am sorry.